src/Algorithm.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace Kmeans;
+
+use Kmeans\Cluster;
+use Kmeans\ClusterCollection;
+use Kmeans\Interfaces\AlgorithmInterface;
+use Kmeans\Interfaces\ClusterCollectionInterface;
+use Kmeans\Interfaces\ClusterInterface;
+use Kmeans\Interfaces\InitializationSchemeInterface;
+use Kmeans\Interfaces\PointCollectionInterface;
+use Kmeans\Interfaces\PointInterface;
+
+class Algorithm implements AlgorithmInterface
+{
+    private InitializationSchemeInterface $initScheme;
+    /** @var array<callable> */
+    private array $iterationCallbacks = [];
+
+
+    public function __construct(InitializationSchemeInterface $initScheme)
+    {
+        $this->initScheme = $initScheme;
+    }
+
+    public function registerIterationCallback(callable $callback): void
+    {
+        $this->iterationCallbacks[] = $callback;
+    }
+
+    public function clusterize(PointCollectionInterface $points, int $nbClusters): ClusterCollectionInterface
```

Comment:
https://github.com/bdelespierre/php-kmeans/issues/28#issuecomment-913543885

The return of Algorithm::clusterize is discussed as ClusterizationResultInterface .

Reply:
Yes, in another PR +1